Sming/Core/Network/SmtpClient.cpp

```diff
@@ -143,45 +139,39 @@ void SmtpClient::onReadyToSendData(TcpConnectionEvent sourceEvent)
 	}
 
 	case eSMTP_SendAuth: {
-		if(authMethods.count()) {
-			auto methodPlain = F("PLAIN");
-			auto methodCramMd5 = F("CRAM-MD5");
-			// TODO: Simplify the code in that block...
-			Vector<String> preferredOrder;
-			if(useSsl) {
-				preferredOrder.addElement(methodPlain);
-				preferredOrder.addElement(methodCramMd5);
-			} else {
-				preferredOrder.addElement(methodCramMd5);
-				preferredOrder.addElement(methodPlain);
-			}
-
-			for(unsigned i = 0; i < preferredOrder.count(); i++) {
-				if(authMethods.contains(preferredOrder[i])) {
-					if(preferredOrder[i] == methodPlain) {
-						// base64('\0' + username + '\0' + password)
-						String token = '\0' + url.User + '\0' + url.Password;
-						String hash = base64_encode(token);
-						sendString(F("AUTH PLAIN ") + hash + "\r\n");
-						state = eSMTP_SendingAuth;
-						break;
-					} else if(preferredOrder[i] == methodCramMd5) {
-						// otherwise we can try the slow cram-md5 authentication...
-						sendString(F("AUTH CRAM-MD5\r\n"));
-						state = eSMTP_RequestingAuthChallenge;
-						break;
-					}
-				}
+		auto authPlain = [this]() {
+			// base64('\0' + username + '\0' + password)
+			String token = '\0' + url.User + '\0' + url.Password;
+			String hash = base64_encode(token);
+			sendString(F("AUTH PLAIN ") + hash + "\r\n");
+			state = eSMTP_SendingAuth;
+		};
+
+		auto authCramMd5 = [this]() {
+			// Slower cram-md5 authentication
+			sendString(F("AUTH CRAM-MD5\r\n"));
+			state = eSMTP_RequestingAuthChallenge;
+		};
+
+		DEFINE_FSTR_LOCAL(methodPlain, "PLAIN")
+		DEFINE_FSTR_LOCAL(methodCramMd5, "CRAM-MD5")
+		if(useSsl) {
+			if(authMethods.contains(methodPlain)) {
+				authPlain();
+			} else if(authMethods.contains(methodCramMd5)) {
+				authCramMd5();
 			}
-		} /* authMethods.count */
+		} else if(authMethods.contains(methodCramMd5)) {
+			authCramMd5();
+		} else if(authMethods.contains(methodPlain)) {
+			authPlain();
+		}
 
 		if(state == eSMTP_SendAuth) {
 			state = eSMTP_Ready;
 		}
 
 		break;
-
-	default:; // Do nothing
```

Comment:
It's a bit weird having this in the middle of the switch statement - I moved it to the normal place. In fact, This appears to be an error as it would handled all other states as if they were `eSMTP_SendAuthResponse`. I've re-tested develop branch and it fails the transfer with various messages, ending in `UNSUPPORTED_VERSION`. With this change it all works as expected.

Reply:
Interesting. I have to see how the `default` ended in the wrong place. Thanks for pointing this out.